src/category_theory/endofunctor/algebra.lean

```diff
@@ -194,5 +201,245 @@ lemma str_is_iso (h : limits.is_initial A) : is_iso A.str :=
+{ obj := λ A, A.1,
+  map := λ A B f, f.1 }
+
+/-- A coalgebra morphism with an underlying isomorphism hom in `C` is a coalgebra isomorphism. -/
+lemma iso_of_iso (f : V₀ ⟶ V₁) [is_iso f.1] : is_iso f :=
+⟨⟨{ f := inv f.1,
+    h' := by { rw [is_iso.eq_inv_comp f.1, ← category.assoc, ← f.h, category.assoc], simp } },
+          by tidy⟩⟩
+
+instance forget_reflects_iso : reflects_isomorphisms (forget F) :=
+{ reflects := λ A B, iso_of_iso }
+
+instance forget_faithful : faithful (forget F) := {}
+
+/--
+From a natural transformation `α : F → G` we get a functor from
+coalgebras of `F` to coalgebras of `G`.
+-/
+@[simps]
+def functor_of_nat_trans {F G : C ⥤ C} (α : F ⟶ G) : coalgebra F ⥤ coalgebra G :=
+{ obj := λ V,
+  { V := V.1,
+    str := V.str ≫ α.app V.1 },
+  map := λ V₀ V₁ f, { f := f.1,
+      h' := begin rw [category.assoc, ← α.naturality, ← category.assoc, f.h, category.assoc] end } }
+
+/-- The identity transformation induces the identity endofunctor on the category of coalgebras. -/
+@[simps {rhs_md := semireducible}]
+def functor_of_nat_trans_id :
+  functor_of_nat_trans (𝟙 F) ≅ 𝟭 _ :=
+nat_iso.of_components
+  (λ X, iso_mk (iso.refl _) (by { dsimp, simp, }))
+  (λ X Y f, by { ext, dsimp, simp })
+
+/-- A composition of natural transformations gives the composition of corresponding functors. -/
+@[simps {rhs_md := semireducible}]
+def functor_of_nat_trans_comp {F₀ F₁ F₂ : C ⥤ C} (α : F₀ ⟶ F₁) (β : F₁ ⟶ F₂) :
+  functor_of_nat_trans (α ≫ β) ≅
+    functor_of_nat_trans α ⋙ functor_of_nat_trans β :=
+nat_iso.of_components
+  (λ X, iso_mk (iso.refl _) (by { dsimp, simp }))
+  (λ X Y f, by { ext, dsimp, simp })
+
+/--
+If `α` and `β` are two equal natural transformations, then the functors of coalgebras induced by
+them are isomorphic.
+We define it like this as opposed to using `eq_to_iso` so that the components are nicer to prove
+lemmas about.
+-/
+@[simps {rhs_md := semireducible}]
+def functor_of_nat_trans_eq {F G : C ⥤ C} {α β : F ⟶ G} (h : α = β) :
+  functor_of_nat_trans α ≅ functor_of_nat_trans β :=
+nat_iso.of_components
+  (λ X, iso_mk (iso.refl _) (by { dsimp, simp [h] }))
+  (λ X Y f, by { ext, dsimp, simp })
+
+/--
+Naturally isomorphic endofunctors give equivalent categories of coalgebras.
+Furthermore, they are equivalent as categories over `C`, that is,
+we have `equiv_of_nat_iso h ⋙ forget = forget`.
+-/
+@[simps]
+def equiv_of_nat_iso {F G : C ⥤ C} (α : F ≅ G) :
+  coalgebra F ≌ coalgebra G :=
+{ functor := functor_of_nat_trans α.hom,
+  inverse := functor_of_nat_trans α.inv,
+  unit_iso :=
+    functor_of_nat_trans_id.symm ≪≫
+    functor_of_nat_trans_eq (by simp) ≪≫
+    functor_of_nat_trans_comp _ _,
+  counit_iso :=
+    (functor_of_nat_trans_comp _ _).symm ≪≫
+    functor_of_nat_trans_eq (by simp) ≪≫
+    functor_of_nat_trans_id }.
+
+end coalgebra
+
+namespace adjunction
+
+variables {F : C ⥤ C} {G : C ⥤ C}
+
+lemma adj_hom_equiv_comp_G_f_eq_f_comp_adj_hom_equiv (adj : F ⊣ G) (A₁ A₂ : algebra F)
```

Comment:
would something like `hom_equiv_str_comp_right_adjoint_map_endofunctor_hom_eq` be a better name? I think this would be a lemma that might come up a lot when working with adjunctions of endofunctors. (I didn't keep the `adj` since the namespace includes `adjunction`, what do you think?

Reply:
Noticing a different proof, I changed to something different, let me know what you think.